File: source/c/sqlite_crud.c
// gcc -o sqlite_crud sqlite_crud.c -lsqlite3
// Requires: libsqlite3-dev
#include <stdio.h>
#include <string.h>
#include <sqlite3.h>

sqlite3 *db;

void exec_sql(const char *sql) {
    char *err = NULL;
    sqlite3_exec(db, sql, NULL, NULL, &err);
    if (err) { fprintf(stderr, "SQL error: %s\n", err); sqlite3_free(err); }
}

int main(void) {
    sqlite3_open(":memory:", &db);

    exec_sql("CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT, email TEXT, age INTEGER)");
    printf("=== Created users table ===\n");

    sqlite3_stmt *stmt;
    const char *insert_sql = "INSERT INTO users (name, email, age) VALUES (?, ?, ?)";

    typedef struct { const char *name; const char *email; int age; } UserRow;
    UserRow inserts[] = {
        {"Alice", "alice@example.com", 30},
        {"Bob", "bob@example.com", 25},
        {"Charlie", "charlie@example.com", 35},
        {"Diana", "diana@example.com", 28}
    };
    for (int i = 0; i < 4; i++) {
        sqlite3_prepare_v2(db, insert_sql, -1, &stmt, NULL);
        sqlite3_bind_text(stmt, 1, inserts[i].name, -1, SQLITE_STATIC);
        sqlite3_bind_text(stmt, 2, inserts[i].email, -1, SQLITE_STATIC);
        sqlite3_bind_int(stmt, 3, inserts[i].age);
        sqlite3_step(stmt);
        sqlite3_finalize(stmt);
    }
    printf("Inserted 4 users\n");

    printf("\n=== All Users ===\n");
    sqlite3_prepare_v2(db, "SELECT * FROM users", -1, &stmt, NULL);
    while (sqlite3_step(stmt) == SQLITE_ROW) {
        printf("  id=%d name=%s email=%s age=%d\n",
               sqlite3_column_int(stmt, 0),
               sqlite3_column_text(stmt, 1),
               sqlite3_column_text(stmt, 2),
               sqlite3_column_int(stmt, 3));
    }
    sqlite3_finalize(stmt);

    printf("\n=== Users over 28 ===\n");
    sqlite3_prepare_v2(db, "SELECT name, age FROM users WHERE age > ?", -1, &stmt, NULL);
    sqlite3_bind_int(stmt, 1, 28);
    while (sqlite3_step(stmt) == SQLITE_ROW) {
        printf("  %s (age %d)\n", sqlite3_column_text(stmt, 0), sqlite3_column_int(stmt, 1));
    }
    sqlite3_finalize(stmt);

    sqlite3_prepare_v2(db, "UPDATE users SET age = 31 WHERE name = ?", -1, &stmt, NULL);
    sqlite3_bind_text(stmt, 1, "Alice", -1, SQLITE_STATIC);
    sqlite3_step(stmt);
    sqlite3_finalize(stmt);
    int changed = sqlite3_changes(db);
    printf("\n=== Updated %d row(s) ===\n", changed);

    sqlite3_prepare_v2(db, "SELECT name, age FROM users WHERE name = ?", -1, &stmt, NULL);
    sqlite3_bind_text(stmt, 1, "Alice", -1, SQLITE_STATIC);
    sqlite3_step(stmt);
    printf("  Alice is now age %d\n", sqlite3_column_int(stmt, 1));
    sqlite3_finalize(stmt);

    sqlite3_prepare_v2(db, "DELETE FROM users WHERE name = ?", -1, &stmt, NULL);
    sqlite3_bind_text(stmt, 1, "Charlie", -1, SQLITE_STATIC);
    sqlite3_step(stmt);
    sqlite3_finalize(stmt);
    int deleted = sqlite3_changes(db);
    printf("\n=== Deleted %d row(s) ===\n", deleted);

    sqlite3_prepare_v2(db, "SELECT name FROM users", -1, &stmt, NULL);
    int first = 1;
    printf("  Remaining users: ");
    while (sqlite3_step(stmt) == SQLITE_ROW) {
        if (!first) printf(", ");
        printf("%s", sqlite3_column_text(stmt, 0));
        first = 0;
    }
    printf("\n");
    sqlite3_finalize(stmt);

    sqlite3_prepare_v2(db, "SELECT COUNT(*) as cnt FROM users", -1, &stmt, NULL);
    sqlite3_step(stmt);
    printf("  Total users: %d\n", sqlite3_column_int(stmt, 0));
    sqlite3_finalize(stmt);

    sqlite3_close(db);
    printf("\nDatabase closed.\n");
    return 0;
}

File: source/javascript/sqlite_crud.js
// SQLite CRUD operations using better-sqlite3 if available, otherwise simulate
let db;
try {
    const Database = require('better-sqlite3');
    db = new Database(':memory:');

    db.exec("CREATE TABLE users (id INTEGER PRIMARY KEY, name TEXT, email TEXT, age INTEGER)");
    console.log("=== Created users table ===");

    const insert = db.prepare("INSERT INTO users (name, email, age) VALUES (?, ?, ?)");
    insert.run("Alice", "alice@example.com", 30);
    insert.run("Bob", "bob@example.com", 25);
    insert.run("Charlie", "charlie@example.com", 35);
    insert.run("Diana", "diana@example.com", 28);
    console.log("Inserted 4 users");

    console.log("\n=== All Users ===");
    const rows = db.prepare("SELECT * FROM users").all();
    for (const row of rows) console.log(`  id=${row.id} name=${row.name} email=${row.email} age=${row.age}`);

    console.log("\n=== Users over 28 ===");
    const older = db.prepare("SELECT name, age FROM users WHERE age > ?").all(28);
    for (const row of older) console.log(`  ${row.name} (age ${row.age})`);

    const changed = db.prepare("UPDATE users SET age = 31 WHERE name = ?").run("Alice").changes;
    console.log(`\n=== Updated ${changed} row(s) ===`);
    const alice = db.prepare("SELECT name, age FROM users WHERE name = ?").all("Alice");
    console.log(`  Alice is now age ${alice[0].age}`);

    const deleted = db.prepare("DELETE FROM users WHERE name = ?").run("Charlie").changes;
    console.log(`\n=== Deleted ${deleted} row(s) ===`);
    const remaining = db.prepare("SELECT name FROM users").all();
    console.log(`  Remaining users: ${remaining.map(r => r.name).join(", ")}`);

    const countRows = db.prepare("SELECT COUNT(*) as cnt FROM users").all();
    console.log(`  Total users: ${countRows[0].cnt}`);

    db.close();
    console.log("\nDatabase closed.");
} catch (e) {
    // Simulate without real SQLite
    console.log("=== Created users table ===");
    let users = [];
    let nextId = 1;

    function insertUser(name, email, age) { users.push({ id: nextId++, name, email, age }); }
    insertUser("Alice", "alice@example.com", 30);
    insertUser("Bob", "bob@example.com", 25);
    insertUser("Charlie", "charlie@example.com", 35);
    insertUser("Diana", "diana@example.com", 28);
    console.log("Inserted 4 users");

    console.log("\n=== All Users ===");
    for (const row of users) console.log(`  id=${row.id} name=${row.name} email=${row.email} age=${row.age}`);

    console.log("\n=== Users over 28 ===");
    for (const row of users.filter(u => u.age > 28)) console.log(`  ${row.name} (age ${row.age})`);

    let changed = 0;
    for (const u of users) { if (u.name === "Alice") { u.age = 31; changed++; } }
    console.log(`\n=== Updated ${changed} row(s) ===`);
    console.log(`  Alice is now age ${users.find(u => u.name === "Alice").age}`);

    const beforeLen = users.length;
    users = users.filter(u => u.name !== "Charlie");
    const deleted = beforeLen - users.length;
    console.log(`\n=== Deleted ${deleted} row(s) ===`);
    console.log(`  Remaining users: ${users.map(u => u.name).join(", ")}`);
    console.log(`  Total users: ${users.length}`);

    console.log("\nDatabase closed.");
}
